Feature/chromiumembeddedframework

ResearchDaniel opened the pull request:
Web browser support. Tested on Mac and Windows 
Keyboard keys not correctly mapped yet.

ResearchDaniel committed:
ChromiumEmbeddedFramework: Initial module structure. Compiles but needs testing

ResearchDaniel committed:
Chromium: Updated to changes in core. Work on getting it up and running

ResearchDaniel committed:
Fix for compiling with UNICODE characters on windows

ResearchDaniel committed:
WebBrowser: Added working module for rendering webpages

ResearchDaniel committed:
Removed old chromiumembeddedframework module, renamed to webbrowser

ResearchDaniel committed:
WebBrowser: Registered CEF license

ResearchDaniel committed:
Webbrowser: Download CEF into build/ext directory

ResearchDaniel committed:
Webbrowser: Formating and review comments

ResearchDaniel committed:
Webbrowser: set CEF_SANBOX to false

ResearchDaniel committed:
Webbrowser: Removed warnings and XCode compile error.

ResearchDaniel committed:
Webbrowser: Adaptions for MacOSX. Fixed dangling pointers

ResearchDaniel committed:
Webbrowser: Minor fix for web helper executable extension. Mostly formating

ResearchDaniel committed:
Webbrowser: Licence year update

r-englund commented:
Works on my machine (windows msvc 2017) 
![image](https://user-images.githubusercontent.com/534670/36483199-c70e5cda-1715-11e8-89e8-0b05737eec14.png)

r-englund commented:
Still get CMake warning
```
CMake Warning at D:/inviwo/builds/web/ext/cef/cef_binary_3.3202.1680.g700054b_windows64/cmake/cef_variables.cmake:342 (message):
  CEF sandbox is not compatible with the current MSVC version (1912)
Call Stack (most recent call first):
  D:/inviwo/builds/web/ext/cef/cef_binary_3.3202.1680.g700054b_windows64/cmake/FindCEF.cmake:38 (include)
  modules/webbrowser/CMakeLists.txt:27 (find_package)
```
Will disapear if  `option(USE_SANDBOX "Enable or disable use of the sandbox." OFF)` is added before the find_package. **NOTE: ** adding `set(USE_SANDBOX FALSE)` is not enough for warning to disappear

r-englund commented:
Crashes in debug mode on windows: 
```
[0221/171601.708:FATAL:dwrite_font_proxy_init_win.cc(95)] Check failed: fallback_available == base::win::GetVersion() > base::win::VERSION_WIN8 (1 vs. 0)
[0221/171601.940:FATAL:dwrite_font_proxy_init_win.cc(95)] Check failed: fallback_available == base::win::GetVersion() > base::win::VERSION_WIN8 (1 vs. 0)
```

ResearchDaniel commented:
@r-englund Win8?

r-englund commented:
Dont know, I have Windows 10 :) 
And it worked fine in RelWithDbgInfo

ResearchDaniel committed:
Webbrowser: Fix error on Windows with debug mode: "FATAL:dwrite_font_proxy_init_win.cc(88)] Check failed: fallback_available == base::win::GetVersion() > base::win::VERSION_WIN8 (1 vs. 0)"

martinfalk commented on cmake/compileoptions.cmake:
```diff
@@ -113,6 +113,7 @@ macro(ivw_define_standard_definitions project_name target)
                                                      NOMINMAX
                                                      WIN32_LEAN_AND_MEAN
                                                      UNICODE
+													 _UNICODE
```
tabs are evil :)

martinfalk commented on modules/webbrowser/CMakeLists.txt:
```diff
@@ -0,0 +1,220 @@
+    COPY_FILES("inviwo-module-webbrowser" "${CEF_BINARY_FILES}" "${CEF_BINARY_DIR}" "${CEF_TARGET_OUT_DIR}")
+    COPY_FILES("inviwo-module-webbrowser" "${CEF_RESOURCE_FILES}" "${CEF_RESOURCE_DIR}" "${CEF_TARGET_OUT_DIR}")
+endif()
+
```
have you tried the packaging and checked whether the zip file and the installer come with the necessary files?

martinfalk commented on modules/webbrowser/interaction/cefinteractionhandler.cpp:
```diff
@@ -0,0 +1,131 @@
+                                       static_cast<int>(wheelEvent->delta().y));
+            event->markAsUsed();
+            break;
+        }
```
what about touch events?

ResearchDaniel replied:
Not supported by Cef when using offscreen rendering to my knowledge. Perhaps we can inject them as mouse events anyway

martinfalk commented on modules/webbrowser/cefprocesshelper.cpp:
```diff
@@ -0,0 +1,59 @@
+ *
+ *********************************************************************************/
+
+#include "modules/webbrowser/webbrowserapp.h"
```
should be a global include

martinfalk commented on modules/webbrowser/interaction/cefinteractionhandler.cpp:
```diff
@@ -0,0 +1,131 @@
+    // https://msdn.microsoft.com/en-us/library/ms646286(VS.85).aspx
+    cefEvent.is_system_key = (e->key() == IvwKey::F10) || (e->modifiers() & KeyModifier::Alt);
+#else
+    // Always forle on non-windows platforms
```
`forle`?

ResearchDaniel replied:
Good question, guessing false :D

martinfalk commented on modules/webbrowser/interaction/cefkeyboardmapping.cpp:
```diff
@@ -0,0 +1,151 @@
+        case IvwKey::Delete:
+            return 46;
+        case IvwKey::A:
+            return 97;
```
what is the reason for capturing `a` and returning a lowercase `a`?
why not return `return 'a';` ?

ResearchDaniel replied:
Cef keycodes are different. We need to investigate more because they don't translate properly at the moment. Also, we need to revamp our KeyEvent to capture more keys.

martinfalk commented on modules/webbrowser/interaction/cefkeyboardmapping.cpp:
```diff
@@ -0,0 +1,151 @@
+        case IvwKey::NumLock:
+        case IvwKey::PrintScreen:
+        case IvwKey::Pause:
+            cefModifiers |= EVENTFLAG_IS_KEY_PAD;
```
for the pause key?

ResearchDaniel replied:
Removed Pause,ScrollLock, PrintScreen, CapsLock. 
Some of these might not be keypad keys either (left, right) depending on which ones it is. Not sure if we currently detect this

martinfalk replied:
At the moment we do no distinguish between regular and numpad keys. Qt provides a key modifier for numpad keys though.

martinfalk commented on modules/webbrowser/interaction/cefkeyboardmapping.cpp:
```diff
@@ -0,0 +1,151 @@
+            break;
+        case IvwKey::LeftMeta:
+        case IvwKey::RightMeta:
+            cefModifiers |= EVENTFLAG_IS_KEY_PAD;
```
why keypad?

ResearchDaniel replied:
I think that I followed this:
https://github.com/bkeiren/cef/blob/master/tests/cefclient/browser/util_win.cc

martinfalk commented on modules/webbrowser/processors/webbrowserprocessor.cpp:
```diff
@@ -0,0 +1,130 @@
+    "Web browser",            // Display name
+    "Web",                    // Category
+    CodeState::Experimental,  // Code state
+    Tags::GL,                 // Tags
```
more flags? E.g. `"GL, Web Browser"`

martinfalk commented on modules/webbrowser/processors/webbrowserprocessor.cpp:
```diff
@@ -0,0 +1,130 @@
+
+        CefBrowserSettings browserSettings;
+
+        browserSettings.windowless_frame_rate = 30;  // 30 is default
```
should this be a system setting instead?

martinfalk commented on modules/webbrowser/renderhandlergl.cpp:
```diff
@@ -0,0 +1,100 @@
+        //        auto index = row * width + col;
+        //        auto outIndex = width * (height - row - 1) + col;
+        //        const auto& pixel = indataV[index];
+        //        // Swizzle, bgra -> rgba
```
remove code?

ResearchDaniel replied:
Kept it for reference if one would like to compare GPU/RAM later

martinfalk commented on modules/webbrowser/renderhandlergl.h:
```diff
@@ -0,0 +1,78 @@
+    RenderHandlerGL(std::function<void()> onWebPageCopiedCallback);
+    void updateCanvasSize(size2_t newSize);
+
+    bool GetViewRect(CefRefPtr<CefBrowser> browser, CefRect &rect) override;
```
any reason for uppercase function names?

ResearchDaniel replied:
Cef standard (overriding their function)

martinfalk commented on modules/webbrowser/webbrowserapp-cotire-fix.cpp:
```diff
@@ -0,0 +1,46 @@
+//    //CefRefPtr<CefV8Value> val = CefV8Value::CreateBool(true);
+//    //CefRefPtr<CefV8Value> global = context->GetGlobal();
+//    //global->SetValue("ExampleValue", val, V8_PROPERTY_ATTRIBUTE_NONE);
+//}
```
remove?

martinfalk commented on modules/webbrowser/webbrowserapp-cotire-fix.h:
```diff
@@ -0,0 +1,56 @@
+    WebBrowserApp();
+
+    CefRefPtr<CefRenderProcessHandler> GetRenderProcessHandler();
+    // void OnContextCreated(CefRefPtr<CefBrowser>, CefRefPtr<CefFrame>, CefRefPtr<CefV8Context>);
```
remove?

martinfalk commented on modules/webbrowser/webbrowserapp.cpp:
```diff
@@ -0,0 +1,46 @@
+//    //CefRefPtr<CefV8Value> val = CefV8Value::CreateBool(true);
+//    //CefRefPtr<CefV8Value> global = context->GetGlobal();
+//    //global->SetValue("ExampleValue", val, V8_PROPERTY_ATTRIBUTE_NONE);
+//}
```
remove?

martinfalk commented on modules/webbrowser/webbrowserapp.h:
```diff
@@ -0,0 +1,59 @@
+    WebBrowserApp();
+
+    CefRefPtr<CefRenderProcessHandler> GetRenderProcessHandler() override;
+    // void OnContextCreated(CefRefPtr<CefBrowser>, CefRefPtr<CefFrame>, CefRefPtr<CefV8Context>);
```
remove

martinfalk commented on modules/webbrowser/webbrowsermodule.cpp:
```diff
@@ -0,0 +1,130 @@
+                    std::string("/../Frameworks/Chromium Embedded Framework.framework/Resources"))
+                       .c_str());
+
+    // Locale returns "en_US.UFT8" but "en.UTF8" is needed by CEF
```
is Unix not affected by this as well?

ResearchDaniel replied:
Not sure. Have not been able to test

martinfalk reviewed (changes requested):
nice work. Minor changes necessary (mostly cosmetic).

ResearchDaniel committed:
Webbrowser: Added support for synchronizing Inviwo properties with web browser.

martinfalk commented on modules/webbrowser/cefimageconverter.h:
```diff
@@ -0,0 +1,61 @@
+public:
+    CefImageConverter() = default;
+    
+    void convert(const Texture2D& fromCefOutput, ImageOutport &toInviwOutput);
```
`toInviwOutput`, typo?

martinfalk commented on modules/webbrowser/processors/propertysyncexampleprocessor.cpp:
```diff
@@ -0,0 +1,142 @@
+    "Property Sync Example",                // Display name
+    "Webbrowser",              // Category
+    CodeState::Experimental,  // Code state
+    Tags::None,               // Tags
```
maybe add `Test` as tag

martinfalk commented on modules/webbrowser/processors/webbrowserprocessor.cpp:
```diff
@@ -70,7 +70,7 @@ WebBrowserProcessor::WebBrowserProcessor()
         CefBrowserSettings browserSettings;
 
-        browserSettings.windowless_frame_rate = 30;  // 30 is default
+        browserSettings.windowless_frame_rate = 60;  // 30 is default
```
adjust comment :)

ResearchDaniel committed:
WebBrowser: Minor fixes

martinfalk reviewed (approved):
great work!

ResearchDaniel committed:
Webbrowser: Pass-through mouse press events on transparent parts of webpage. Added support for touch events

ResearchDaniel committed:
WebBrowser:
- Improved performance so it is now acceptable
- Updated CEF lib. Fixes slider issue but crashes on scroll events on Mac
- Now installs files during packaging
- Changed directory to copy CEF framework during build so that it works for more than inviwo network editor
- CEF scroll events are only sent if mouse is over non-transparent pixel
- Added optional background to processors

ResearchDaniel committed:
WebBrowser: Using picking events instead of checking transparent pixels.

ResearchDaniel committed:
WebBrowser: TouchEvent fix

ResearchDaniel committed:
Merge branch 'master' into feature/chromiumembeddedframework

ResearchDaniel committed:
WebBrowser: Refactoring

ResearchDaniel committed:
WebBrowser: Copyright year

ResearchDaniel committed:
Added CanvasHideEvent and CanvasShowEvent to be able to stop web browser rendering when canvas is not visible. These two events are propagated before hiding and showing the CanvasProcessorWidget.

ResearchDaniel committed:
Webbrowser: Check if element is found in html before using it (crashed otherwise)

ResearchDaniel committed:
Webbrowser: Webbrowser can no dynamically add properties to sync. Changed synchronization to be done through JSON instead of Serialization since serialization caused trouble with unspecified values such as identifiers.

petersteneteg commented on modules/webbrowser/interaction/cefinteractionhandler.cpp:
```diff
@@ -52,6 +54,18 @@ void CEFInteractionHandler::invokeEvent(Event* event) {
             event->markAsUsed();
             break;
         }
+        case CanvasHideEvent::chash(): {
```
I don't think this is a good way of doing this, what if more then one canvas is attached? what about other types of sinks?
Should really check if there is a connected sink... only the networkevaluator really knows this right now. But I you want we might be able to expose that...

ResearchDaniel replied:
Good point! Better to use onProcessorReadyChanged?

petersteneteg replied:
What you could do now is to add your self as an observer to the processor network on processor added removed.
then you would iterate over all you out connected processors add your self as observer to them and override the onProcessorSinkChanged, then you can respond whenever the sink status of your dependent processors change, which determines whether you should be evaluating or not...

petersteneteg replied:
If you want I can explain further in person tomorrow. We chould also discuss if the "will be evaluated ness" of a processor should/needs to be exposed in a easier way...

ResearchDaniel committed:
Webbrowser: Got it to work with newest CEF version. Crashed when settings.windowless_rendering_enabled==false. Could also be a combined issue that the CefBrowserProcess was mixed up with the CefRenderProcess. No issues with scrolling anymore
- Updated to newest CEF version
- RendererApp for web helper
- BrowserApp for Inviwo-process
- Enabled offscreen rendering setting

petersteneteg commented:
Think it would be nice to add a minumum app using Cef also...

ResearchDaniel commented:
@petersteneteg Probably quite some work since it would involve using native window handles and converting and forwarding native events. Perhaps a canvas processor could be used for this purpose in that case...

petersteneteg commented:
Maybe use glfw our side?

ResearchDaniel committed:
Core: Removed canvas show/hide event again since it does not work with multiple connected canvas’

ResearchDaniel committed:
Core/GLFW: Added unicode text to KeyboardEvent to be able to write text. Adapted CanvasQt and CanvasGLFW to these changes

ResearchDaniel committed:
Webbrowser: Text input works on Mac. Needs testing on Windows. Fixed crash when default locale ”C” was used

petersteneteg reviewed (commented):
This changes that is not related to CEF should probably go in its own PR...

ResearchDaniel commented:
@petersteneteg Yes, but a bit inconvenient. I think that we can merge this one when it has been tested on Windows. It is good enough for now. Will make it easier to separate changes necessary for it work properly

ResearchDaniel committed:
Core: Changed constructor argument order to make it backwards compatible

petersteneteg commented:
@ResearchDaniel Think we should just merge this now... before it gets old again :)
